### PR TITLE
Gulp

### PR DESCRIPTION
**Pull request by li-kai**
# Changes
- Migration to gulp instead of grunt. Allowing us to move off from callbacks to promises and async/await async.
- Automatically starts scraping data a month before the next semester, so no more manually renaming config.
- examTimetable.json now uses `ModuleCode` instead of `Code` to remain consistent with all other data sources.
- venueInformation.json keys `Day`, `Classes` and `Availability` are now capitalised to be consistent with all other data sources.
- modmaven is rewritten using pure JavaScript which is tested pretty rigorously and has more passing cases.
- API keys are now obtained through environment variables.
- There is logging now and provides much more information than previously.
- A heavy reliance on ramda as an utility library.
- Cache is now organised by root urls, which means we can delete cache files much more easily now.
- Tasks are much more organised and has better documentation.
- Follows airbnb linting.
- A basic skeleton for GraphQL server to be up and running.

Sorry this took so long and it's so big it's like I adopted waterfall instead of being agile. Many parts were half done until the end 😅 and I couldn't PR x as it relied on half done y, and so on.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  "presets": [
+    ["eslatest-node6", { "loose": true }]
+  ],
+  "plugins": [
+    "transform-flow-strip-types",
+    "transform-es2015-destructuring",
+    "transform-es2015-parameters",
+    ["transform-object-rest-spread", { "useBuiltIns": true }]
+  ],
+}
```
Comment (yangshun) [inline .babelrc]: Set your editor to add newline at EOF upon saving

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+root = true
+
+[*]
+indent_style = space
+indent_size = 2
+
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
```
Comment (yangshun) [inline .editorconfig]: This is ironic :joy:

File: __tests__/genReqTree.test.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import parseString from '../gulp-tasks/local/genReqTree/parseString';
```
Comment (yangshun) [inline __tests__/genReqTree.test.js]: Tests! 🎉

File: gulp-tasks/local/consolidateForSem.js [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+    let duration;
+    if (exam.Duration) {
+      duration = `P${exam.Duration.replace(/\s/g, '').toUpperCase().slice(0, 5)}`;
+    }
+    return {
+      ModuleCode: exam.ModuleCode,
+      ExamDate: examString,
+      ExamDuration: duration,
+      ExamOpenBook: exam[''] ? exam[''] === '*' : undefined,
```
Comment (yangshun) [inline gulp-tasks/local/consolidateForSem.js]: I never knew there was this field. Who provides this data?

File: gulp-tasks/local/consolidateForSem.js [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  });
+  return normalizeData(data);
+}
+
+function consolidate(data, subLog) {
+  const mainModuleCodes = [
+    ...Object.keys(data.bulletinModules),
+    ...Object.keys(data.cors),
+  ];
+  const auxilaryModuleCodes = [
```
Comment (yangshun) [inline gulp-tasks/local/consolidateForSem.js]: Nit: auxiliary

File: gulp-tasks/local/genReqTree/index.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+  const filterUnparseable = R.pipe(
+    R.filter(str => !RESTRICTED_KEYWORDS.some(keyword => str.includes(keyword))), // remove restricted
+    R.filter(R.test(MODULE_REGEX)), // remove those with no modules
+  );
+  const parsable = filterUnparseable(moduleCodeToData);
+
+  Object.keys(moduleCodeToData).forEach((moduleCode) => {
+    if (!Object.prototype.hasOwnProperty.call(parsable, moduleCode)) {
+      // log.debug(`${moduleCode}'s ${key} cannot be parsed: ${moduleCodeToData[moduleCode]}`);
```
Comment (yangshun) [inline gulp-tasks/local/genReqTree/index.js]: Is this meant to be commented out?

File: package.json [hunk 7]
```diff
@@ -17,21 +26,63 @@
       "url": "https://github.com/nusmodifications/nusmods-api/blob/master/LICENSE-MIT"
     }
   ],
+  "private": true,
   "dependencies": {
-    "async": "^1.3.0",
+    "babel-cli": "^6.16.0",
```
Comment (yangshun) [inline package.json]: Shall we use yarn?

File: .babelrc (hunk 1 above)
Comment (li-kai) [inline .babelrc]: Hmm I did, nice catch though.

File: gulp-tasks/local/consolidateForSem.js (hunk 4 above)
Comment (li-kai) [inline gulp-tasks/local/consolidateForSem.js]: It's a legacy field now 😭 used to be available in the html files.

File: package.json (hunk 7 above)
Comment (li-kai) [inline package.json]: it's already using yarn!

File: gulp-tasks/local/genReqTree/index.js (hunk 6 above)
Comment (li-kai) [inline gulp-tasks/local/genReqTree/index.js]: Yep. If all debug messages are outputted, the logs just become noise.

**Pull request merged**